connection-loop.ts

```diff
@@ -0,0 +1,262 @@
+let pushCounter = 0;
+const MIN_DELAY = 30;
+const MAX_DELAY = 60_000;
+
+const MAX_CONNECTIONS = 3;
+
+// Number of connections to remember when computing the new delay.
+const CONNECTION_MEMORY_COUNT = 10;
+
+type Duration = {start: number; stop: number; ok: boolean};
+type Durations = Duration[];
+
+const PUSH_DELAY = 10;
+
+class Replicache {
+  private _pendingPushResolver = resolver<void>();
+  private _currentPushResolver = resolver<void>();
+
+  async push() {
+    this._pendingPushResolver.resolve();
+    // await a promise to allow the currentPushResolver to be created inside the
+    // connection loop.
+    await Promise.resolve(0);
+    await this._currentPushResolver.promise;
+  }
+
+  async connectionLoop() {
+    const durations: Durations = [];
+    let delay = MIN_DELAY;
+    let lastConnectionErrored = false;
+    console.log('Starting connection loop');
+    for (;;) {
+      // SHould be while (!closed)?
+      console.log('Waiting for a push');
+      await this._pendingPushResolver.promise;
+      this._currentPushResolver = resolver();
```

Comment:
It seems like this can overwrite the value that will later be used in line 84, which I guess would mean that a call to `push()` might not resolve?

Reply:
Did you consider tracking the set of waiting pushes and resolving them explicitly?

Imagining creating a resolver in push, storing that in an array, then await the corresponding promise. Later, when we actually send the push asynchronously, we can pass that array into the async operation, and resolve all the waiting calls explicitly.

Reply:
I think you are right. Will look at it

Reply:
A simpler solution is to use a local variable.

Reply:
Looks like using a local variable should be enough after all...

Reply:
Hm, will be curious to see how this would work.